File: go/indicators/welleswilder/averagedirectionalmovementindex/averagedirectionalmovementindex.go
package averagedirectionalmovementindex

import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/core"
	"zpano/indicators/welleswilder/directionalmovementindex"
)

const (
	adxMnemonic    = "adx"
	adxDescription = "Average Directional Movement Index"
)

// AverageDirectionalMovementIndex is Welles Wilder's Average Directional Movement Index (ADX).
//
// The average directional movement index smooths the directional movement index (DX)
// using Wilder's smoothing technique. It is calculated as:
//
//	Initial ADX = SMA of first `length` DX values
//	Subsequent ADX = (previousADX * (length-1) + DX) / length
//
// The indicator requires close, high, and low values.
//
// Reference:
//
// Wilder, J. Welles Jr. (1978). New Concepts in Technical Trading Systems.
type AverageDirectionalMovementIndex struct {
	mu                       sync.RWMutex
	length                   int
	lengthMinusOne           float64
	count                    int
	sum                      float64
	primed                   bool
	value                    float64
	directionalMovementIndex *directionalmovementindex.DirectionalMovementIndex
}

// NewAverageDirectionalMovementIndex returns a new instance of the Average Directional Movement Index indicator.
func NewAverageDirectionalMovementIndex(length int) (*AverageDirectionalMovementIndex, error) {
	if length < 1 {
		return nil, fmt.Errorf("invalid length %d: must be >= 1", length)
	}

	dx, err := directionalmovementindex.NewDirectionalMovementIndex(length)
	if err != nil {
		return nil, fmt.Errorf("failed to create directional movement index: %w", err)
	}

	return &AverageDirectionalMovementIndex{
		length:                   length,
		lengthMinusOne:           float64(length - 1),
		value:                    math.NaN(),
		directionalMovementIndex: dx,
	}, nil
}

// Length returns the length parameter.
func (s *AverageDirectionalMovementIndex) Length() int {
	return s.length
}

// IsPrimed indicates whether the indicator is primed.
func (s *AverageDirectionalMovementIndex) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes the output data of the indicator.
func (s *AverageDirectionalMovementIndex) Metadata() core.Metadata {
	return core.BuildMetadata(
		core.AverageDirectionalMovementIndex,
		adxMnemonic,
		adxDescription,
		[]core.OutputText{
			{Mnemonic: adxMnemonic, Description: adxDescription},
			{Mnemonic: "dx", Description: "Directional Movement Index"},
			{Mnemonic: "+di", Description: "Directional Indicator Plus"},
			{Mnemonic: "-di", Description: "Directional Indicator Minus"},
			{Mnemonic: "+dm", Description: "Directional Movement Plus"},
			{Mnemonic: "-dm", Description: "Directional Movement Minus"},
			{Mnemonic: "atr", Description: "Average True Range"},
			{Mnemonic: "tr", Description: "True Range"},
		},
	)
}

// Update updates the Average Directional Movement Index given the next bar's close, high, and low values.
func (s *AverageDirectionalMovementIndex) Update(close, high, low float64) float64 {
	if math.IsNaN(close) || math.IsNaN(high) || math.IsNaN(low) {
		return math.NaN()
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	dxValue := s.directionalMovementIndex.Update(close, high, low)

	if !s.directionalMovementIndex.IsPrimed() {
		return math.NaN()
	}

	if s.primed {
		s.value = (s.value*s.lengthMinusOne + dxValue) / float64(s.length)
		return s.value
	}

	s.count++
	s.sum += dxValue

	if s.count == s.length {
		s.value = s.sum / float64(s.length)
		s.primed = true

		return s.value
	}

	return math.NaN()
}

// UpdateSample updates the Average Directional Movement Index using a single sample value
// as a substitute for close, high, and low.
func (s *AverageDirectionalMovementIndex) UpdateSample(sample float64) float64 {
	return s.Update(sample, sample, sample)
}

// UpdateScalar updates the indicator given the next scalar sample.
func (s *AverageDirectionalMovementIndex) UpdateScalar(sample *entities.Scalar) core.Output {
	v := sample.Value

	output := make([]any, 1)
	output[0] = entities.Scalar{Time: sample.Time, Value: s.Update(v, v, v)}

	return output
}

// UpdateBar updates the indicator given the next bar sample.
func (s *AverageDirectionalMovementIndex) UpdateBar(sample *entities.Bar) core.Output {
	output := make([]any, 1)
	output[0] = entities.Scalar{Time: sample.Time, Value: s.Update(sample.Close, sample.High, sample.Low)}

	return output
}

// UpdateQuote updates the indicator given the next quote sample.
func (s *AverageDirectionalMovementIndex) UpdateQuote(sample *entities.Quote) core.Output {
	v := (sample.Bid + sample.Ask) / 2 //nolint:mnd

	output := make([]any, 1)
	output[0] = entities.Scalar{Time: sample.Time, Value: s.Update(v, v, v)}

	return output
}

// UpdateTrade updates the indicator given the next trade sample.
func (s *AverageDirectionalMovementIndex) UpdateTrade(sample *entities.Trade) core.Output {
	v := sample.Price

	output := make([]any, 1)
	output[0] = entities.Scalar{Time: sample.Time, Value: s.Update(v, v, v)}

	return output
}

File: ts/indicators/welles-wilder/average-directional-movement-index/average-directional-movement-index.ts
import { buildMetadata } from '../../core/build-metadata';
import { Bar } from '../../../entities/bar';
import { Quote } from '../../../entities/quote';
import { Scalar } from '../../../entities/scalar';
import { Trade } from '../../../entities/trade';
import { Indicator } from '../../core/indicator';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorOutput } from '../../core/indicator-output';
import { IndicatorIdentifier } from '../../core/indicator-identifier';
import { DirectionalMovementIndex } from '../directional-movement-index/directional-movement-index';

const adxMnemonic = 'adx';
const adxDescription = 'Average Directional Movement Index';

/**
 * Welles Wilder's Average Directional Movement Index (ADX).
 *
 * The average directional movement index smooths the directional movement index (DX)
 * using Wilder's smoothing technique. It is calculated as:
 *
 *   Initial ADX = SMA of first `length` DX values
 *   Subsequent ADX = (previousADX * (length-1) + DX) / length
 *
 * The indicator requires close, high, and low values.
 */
export class AverageDirectionalMovementIndex implements Indicator {

  private readonly length_: number;
  private readonly lengthMinusOne: number;
  private count = 0;
  private sum = 0;
  private primed_ = false;
  private value_ = NaN;
  private readonly directionalMovementIndex: DirectionalMovementIndex;

  constructor(length: number) {
    if (length < 1) {
      throw new Error(`invalid length ${length}: must be >= 1`);
    }

    this.length_ = length;
    this.lengthMinusOne = length - 1;
    this.directionalMovementIndex = new DirectionalMovementIndex(length);
  }

  /** Returns the length parameter. */
  public get length(): number {
    return this.length_;
  }

  /** Indicates whether the indicator is primed. */
  public isPrimed(): boolean {
    return this.primed_;
  }

  /** Describes the output data of the indicator. */
  public metadata(): IndicatorMetadata {
    return buildMetadata(
      IndicatorIdentifier.AverageDirectionalMovementIndex,
      adxMnemonic,
      adxDescription,
      [
        { mnemonic: adxMnemonic, description: adxDescription },
        { mnemonic: 'dx', description: 'Directional Movement Index' },
        { mnemonic: '+di', description: 'Directional Indicator Plus' },
        { mnemonic: '-di', description: 'Directional Indicator Minus' },
        { mnemonic: '+dm', description: 'Directional Movement Plus' },
        { mnemonic: '-dm', description: 'Directional Movement Minus' },
        { mnemonic: 'atr', description: 'Average True Range' },
        { mnemonic: 'tr', description: 'True Range' },
      ],
    );
  }

  /** Updates the Average Directional Movement Index given the next bar's close, high, and low values. */
  public update(close: number, high: number, low: number): number {
    if (isNaN(close) || isNaN(high) || isNaN(low)) {
      return NaN;
    }

    const dxValue = this.directionalMovementIndex.update(close, high, low);

    if (!this.directionalMovementIndex.isPrimed()) {
      return NaN;
    }

    if (this.primed_) {
      this.value_ = (this.value_ * this.lengthMinusOne + dxValue) / this.length_;
      return this.value_;
    }

    this.count++;
    this.sum += dxValue;

    if (this.count === this.length_) {
      this.value_ = this.sum / this.length_;
      this.primed_ = true;
      return this.value_;
    }

    return NaN;
  }

  /** Updates the Average Directional Movement Index using a single sample value as a substitute for close, high, and low. */
  public updateSample(sample: number): number {
    return this.update(sample, sample, sample);
  }

  /** Updates the indicator given the next scalar sample. */
  public updateScalar(sample: Scalar): IndicatorOutput {
    const v = sample.value;
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = this.update(v, v, v);
    return [scalar];
  }

  /** Updates the indicator given the next bar sample. */
  public updateBar(sample: Bar): IndicatorOutput {
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = this.update(sample.close, sample.high, sample.low);
    return [scalar];
  }

  /** Updates the indicator given the next quote sample. */
  public updateQuote(sample: Quote): IndicatorOutput {
    const v = (sample.bid + sample.ask) / 2;
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = this.update(v, v, v);
    return [scalar];
  }

  /** Updates the indicator given the next trade sample. */
  public updateTrade(sample: Trade): IndicatorOutput {
    const v = sample.price;
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = this.update(v, v, v);
    return [scalar];
  }
}
